src/tools/non-promotion-reasons.md

```diff
@@ -1,45 +1,399 @@
+doesn't mean the code at (3) is dead;
+`o` might have a type — like `String` —
+that implements both `Comparable` and `Pattern`.
+
+***[PENDING: what's the fix?]***
+
+
+### Write captured by a local function
+
+**The cause:**
+The variable has been write captured by
+a local function or function expression.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(int? i, int? j) {
+  if (i == null) return;
+  var foo = () {
+    i = j;
+  };
+  print(i.isEven); // ERROR
+}
+{% endprettify %}
+
+Flow analysis reasons that as soon as the definition of `foo` is reached,
+it might get called at any time,
+therefore it's no longer safe to promote `i` at all.
+As with loops, this demotion happens regardless of
+the type of the right hand side of the assignment.
+
+***[PENDING: what's the fix?]***
+
+
+### Written outside of the current closure or function expression {#write-outer}
+
+**The cause:**
+The variable is written to outside of a closure or function expression,
+and the type promotion location is
+inside the closure or function expression.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(int? i, int? j) {
+  if (i == null) return;
+  var foo = () {
+    print(i.isEven); // (1) ERROR
+  };
+  i = j;             // (2)
+}
+{% endprettify %}
+
+Flow analysis reasons that there's no way to determine
+when `foo` might get called,
+so it might get called after the assignment at (2),
+and thus the promotion might no longer be valid.
+As with loops, this demotion happens regardless of the type of
+the right hand side of the assignment.
+
+A particularly nasty case looks like this:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(int? i) {
+  i ??= 0;
+  var foo = () {
+    print(i.isEven); // ERROR
+  };
+}
+{% endprettify %}
+
+In this case, a human can see that the promotion is safe because
+the only write to `i` uses a non-null value and
+happens before `foo` is ever created.
+But [flow analysis isn't that smart][1536].
+
+[1536]: https://github.com/dart-lang/language/issues/1536
+
+***[PENDING: what's the fix?]***
+
+
+### Write captured outside of the current closure or function expression {#captured-outer}
+
+**The cause:**
+The variable you're trying to promote is write captured
+outside of a closure or function expression,
+but this use of the variable is inside of the closure or function expression
+that's trying to promote it.
+
+Example:
+
+{:.bad}
+{% prettify dart tag=pre+code %}
+void f(int? i, int? j) {
+  var foo = () {
+    if (i == null) return;
+    print(i.isEven); // ERROR
+  };
+  var bar = () {
+    i = j;
+  };
+}
+{% endprettify %}
+
+Flow analysis reasons that there's no way of telling
+what order `foo` and `bar` might be executed in;
+in fact, `bar` might even get executed halfway through executing `foo`
+(due to `foo` calling something that calls `bar`).
+So it isn't safe to promote `i` at all inside `foo`.
+
+***[PENDING: what's the fix?]***
```

Comment:
The best bet here is probably to do a separate variable, e.g.:
```dart
void f(int? i, int? j) {
  var foo = () {
    var i2 = i;
    if (i2 == null) return;
    print(i2.isEven); // OK; i2 can be promoted because it's local to this closure
  };
  var bar = () {
    i = j;
  };
}
```